strings/base_collections_base.h

```diff
@@ -407,27 +534,36 @@ WINRT_EXPORT namespace winrt
 
         bool Insert(K const& key, V const& value)
         {
-            this->increment_version();
-            auto pair = static_cast<D&>(*this).get_container().insert_or_assign(static_cast<D const&>(*this).wrap_value(key), static_cast<D const&>(*this).wrap_value(value));
-            return !pair.second;
+            return static_cast<D&>(*this).perform_exclusive([&]
+            {
+                this->increment_version();
+                auto pair = static_cast<D&>(*this).get_container().insert_or_assign(static_cast<D const&>(*this).wrap_value(key), static_cast<D const&>(*this).wrap_value(value));
+                return !pair.second;
+            });
         }
 
         void Remove(K const& key)
         {
-            auto& container = static_cast<D&>(*this).get_container();
-            auto found = container.find(static_cast<D const&>(*this).wrap_value(key));
-            if (found == container.end())
+            static_cast<D&>(*this).perform_exclusive([&]
             {
-                throw hresult_out_of_bounds();
-            }
-            this->increment_version();
-            container.erase(found);
+                auto& container = static_cast<D&>(*this).get_container();
+                auto found = container.find(static_cast<D const&>(*this).wrap_value(key));
+                if (found == container.end())
+                {
+                    throw hresult_out_of_bounds();
+                }
+                this->increment_version();
+                container.erase(found);
+            });
         }
 
         void Clear() noexcept
         {
-            this->increment_version();
-            static_cast<D&>(*this).get_container().clear();
+            static_cast<D&>(*this).perform_exclusive([&]
+            {
+                this->increment_version();
+                static_cast<D&>(*this).get_container().clear();
```

Comment:
Need to save the old values so they can be destructed outside the lock. Probably should just swap in an empty container.